File: string/541.反转字符串-ii.go
/*
 * @lc app=leetcode.cn id=541 lang=golang
 *
 * [541] 反转字符串 II
 */

// @lc code=start

func reverseStr(s string, k int) string {
	sb := []byte(s)
	l := len(sb)
	for i := 0; i < l; i += 2 * k {
		if i+k <= l {
			rever(sb[i : i+k])
		} else {
			rever(sb[i:l])
		}
	}
	return string(sb)
}

func rever(b []byte) {
	left := 0
	right := len(b) - 1

	for left <= right {
		b[left], b[right] = b[right], b[left]
		left++
		right--
	}
}

// @lc code=end

File: string/541.反转字符串-ii.py
#
# @lc app=leetcode.cn id=541 lang=python3
#
# [541] 反转字符串 II
#


# @lc code=start
def rever(s: List[str]) -> List[str]:
    left = 0
    right = len(s) - 1
    while left <= right:
        s[left], s[right] = s[right], s[left]
        left += 1
        right -= 1
    return s


class Solution:

    def reverseStr(self, s: str, k: int) -> str:
        l = len(s)
        res = list(s)

        for i in range(0, l, 2 * k):
            res[i : i + k] = rever(res[i : i + k])
        return "".join(res)


# @lc code=end
